## Análisis de Datos con NumPy

### **Objetivo:**

Evaluar las capacidades de los estudiantes en el uso de NumPy para la creación y manipulación de arrays, operaciones matemáticas, indexación, slicing y agregaciones estadísticas. Este ejercicio también busca desarrollar habilidades analíticas y de programación lógica.

### **Instrucciones:**

**Descripción del Problema:** 

Imagina que estás trabajando como analista de datos para una empresa que recopila información sobre las ventas diarias de diferentes productos en varias tiendas. Necesitas analizar estos datos para extraer información útil y presentar un resumen de las ventas.

**Requisitos del Sistema:**

El programa debe generar datos de ventas de manera aleatoria para 10 productos diferentes en 5 tiendas distintas durante 7 días (una semana).

Los datos deben estar almacenados en un array de NumPy de 3 dimensiones (productos, tiendas, días).

**Debes calcular y mostrar la siguiente información:**
- Total de ventas por producto a lo largo de la semana.
- Total de ventas por tienda a lo largo de la semana.
- Promedio de ventas por producto por día.
- Promedio de ventas por tienda por día.
- Producto con mayor y menor ventas totales en la semana.
- Tienda con mayor y menor ventas totales en la semana.

**Detalles Específicos:**
- Genera los datos de ventas aleatorias usando np.random.randint, con valores entre 0 y 100.
- Usa funciones de NumPy para calcular las sumas y promedios.
- Maneja posibles errores o valores inesperados.
- Presenta los resultados de manera clara y estructurada.

**Consideraciones Adicionales:**
- Organiza tu código utilizando funciones para cada una de las tareas principales.
- Usa comentarios para explicar cada sección de tu código.
- Asegúrate de que el código sea robusto y maneje casos en los que los datos generados puedan estar fuera de los rangos esperados.



In [1]:
import numpy as np

# Genera datos de ventas aleatorios para el número de productos, tiendas y días especificados
def generar_datos_ventas(productos, tiendas, dias):
    return np.random.randint(0, 101, (productos, tiendas, dias))  # Rango de 0 a 100, para 10 productos, 5 tiendas y 7 días

# Calcula el total de ventas por producto a lo largo de la semana
def calcular_totales_ventas_por_producto(datos):
    return np.sum(datos, axis=(1, 2))  # Suma a lo largo de tiendas y días para cada producto

# Calcula el total de ventas por tienda a lo largo de la semana
def calcular_totales_ventas_por_tienda(datos):
    return np.sum(datos, axis=(0, 2))  # Suma a lo largo de productos y días para cada tienda

# Calcula el promedio de ventas por producto por día
def calcular_promedio_ventas_por_producto(datos):
    return np.mean(datos, axis=(1, 2))  # Promedio en tiendas y días para cada producto

# Calcula el promedio de ventas por tienda por día
def calcular_promedio_ventas_por_tienda(datos):
    return np.mean(datos, axis=(0, 2))  # Promedio en productos y días para cada tienda

# Encuentra el producto con mayor y menor ventas totales en la semana
def encontrar_producto_mayor_menor_ventas(totales_por_producto):
    producto_mayor_ventas = np.argmax(totales_por_producto)  # Índice del producto con ventas máximas
    producto_menor_ventas = np.argmin(totales_por_producto)  # Índice del producto con ventas mínimas
    return producto_mayor_ventas, producto_menor_ventas

# Encuentra la tienda con mayor y menor ventas totales en la semana
def encontrar_tienda_mayor_menor_ventas(totales_por_tienda):
    tienda_mayor_ventas = np.argmax(totales_por_tienda)  # Índice de la tienda con ventas máximas
    tienda_menor_ventas = np.argmin(totales_por_tienda)  # Índice de la tienda con ventas mínimas
    return tienda_mayor_ventas, tienda_menor_ventas

# Función principal para ejecutar el análisis y mostrar los resultados de manera estructurada
def imprimir_resumen():
    # Configuración inicial de dimensiones
    productos = 10
    tiendas = 5
    dias = 7

    # Genera los datos de ventas
    datos = generar_datos_ventas(productos, tiendas, dias)

    # Calcula los totales y promedios
    totales_por_producto = calcular_totales_ventas_por_producto(datos)
    totales_por_tienda = calcular_totales_ventas_por_tienda(datos)
    promedio_por_producto = calcular_promedio_ventas_por_producto(datos)
    promedio_por_tienda = calcular_promedio_ventas_por_tienda(datos)

    # Encuentra el producto y la tienda con mayor y menor ventas
    producto_mayor_ventas, producto_menor_ventas = encontrar_producto_mayor_menor_ventas(totales_por_producto)
    tienda_mayor_ventas, tienda_menor_ventas = encontrar_tienda_mayor_menor_ventas(totales_por_tienda)

    # Imprime los resultados de manera clara y estructurada
    print("Total de ventas por producto a lo largo de la semana:", totales_por_producto)
    print("Total de ventas por tienda a lo largo de la semana:", totales_por_tienda)
    print("Promedio de ventas por producto por día:", promedio_por_producto)
    print("Promedio de ventas por tienda por día:", promedio_por_tienda)
    print(f"Producto con mayor ventas: Producto {producto_mayor_ventas}")
    print(f"Producto con menor ventas: Producto {producto_menor_ventas}")
    print(f"Tienda con mayor ventas: Tienda {tienda_mayor_ventas}")
    print(f"Tienda con menor ventas: Tienda {tienda_menor_ventas}")

# Ejecuta el análisis de datos de ventas y muestra el resumen
imprimir_resumen()


Total de ventas por producto a lo largo de la semana: [1857 1880 1409 1512 1731 1781 1745 2018 1650 1728]
Total de ventas por tienda a lo largo de la semana: [3763 3115 3585 3571 3277]
Promedio de ventas por producto por día: [53.05714286 53.71428571 40.25714286 43.2        49.45714286 50.88571429
 49.85714286 57.65714286 47.14285714 49.37142857]
Promedio de ventas por tienda por día: [53.75714286 44.5        51.21428571 51.01428571 46.81428571]
Producto con mayor ventas: Producto 7
Producto con menor ventas: Producto 2
Tienda con mayor ventas: Tienda 0
Tienda con menor ventas: Tienda 1


Total de ventas por producto a lo largo de la semana:

- El array [1857, 1880, 1409, 1512, 1731, 1781, 1745, 2018, 1650, 1728] representa el total de ventas de cada uno de los 10 productos a lo largo de la semana. El valor más alto es 2018 (Producto 7) y el más bajo es 1409 (Producto 2).
Total de ventas por tienda a lo largo de la semana:

- El array [3763, 3115, 3585, 3571, 3277] representa el total de ventas para cada una de las 5 tiendas en la semana. La tienda con mayores ventas es la Tienda 0 (3763 ventas), y la tienda con menores ventas es la Tienda 1 (3115 ventas).
Promedio de ventas por producto por día:

- El array [53.05, 53.71, 40.26, 43.2, 49.46, 50.89, 49.86, 57.66, 47.14, 49.37] muestra el promedio diario de ventas para cada producto. Esto te da una idea de cuán constante o variable es la venta diaria de cada producto. Producto 7 tiene el promedio más alto, con 57.66.
Promedio de ventas por tienda por día:

- El array [53.76, 44.5, 51.21, 51.01, 46.81] muestra el promedio diario de ventas por tienda. Esto ayuda a comparar el rendimiento de cada tienda en términos de ventas promedio.
Producto con mayor y menor ventas:

- Producto 7 tiene el mayor total de ventas en la semana, mientras que Producto 2 tiene el menor. Esto permite identificar qué productos son más populares o rentables.
Tienda con mayor y menor ventas:

- Tienda 0 generó el mayor volumen de ventas durante la semana, y Tienda 1 el menor, lo cual es útil para la toma de decisiones sobre distribución y estrategias de marketing específicas para cada tienda.